benchmark/scripts/report.ts

```diff
@@ -0,0 +1,314 @@
+          anchor: "end",
+          clamp: true,
+          formatter: (_, context: Context) =>
+            props.datalabels[context.dataIndex],
+        },
+      },
+      {
+        label: "right",
+        data: props.rightData,
+        backgroundColor: props.backgroundColors,
+        datalabels: {
+          labels: {
+            title: null,
+          },
+        },
+      },
+    ],
+  };
+
+  // Create a new Chart.js instance
+  const width = 800;
+  const height = 600;
+  const backgroundColour = "white";
+  const chartCallback: ChartCallback = (ChartJS) => {
+    ChartJS.defaults.color = "rgba(31, 35, 40)";
+    ChartJS.defaults.font.size = 12;
+    ChartJS.defaults.responsive = false;
+    ChartJS.defaults.maintainAspectRatio = false;
+  };
+  const chartJSNodeCanvas = new ChartJSNodeCanvas({
+    width,
+    height,
+    backgroundColour,
+    plugins: {
+      modern: ["chartjs-plugin-datalabels"],
+    },
+    chartCallback,
+  });
+  const buffer = await chartJSNodeCanvas.renderToBuffer({
+    type: "bar",
+    data,
+    options,
+  });
+  await fsPromises.writeFile(filePath, buffer, "base64");
+};
+
+function generateCharts(
+  datasetType: string,
+  datasets: Record<string, Record<string, any>>,
+  gitBranch: string
+) {
+  for (const key in datasets) {
+    const labels: string[] = [];
+    const datalabels: string[] = [];
+    const backgroundColors: string[] = [];
+    const borderColors: string[] = [];
+    const offsetData: number[] = [];
+    const leftData: number[] = [];
+    const rightData: number[] = [];
+    let max = 0;
+    let min = Infinity;
+    console.log(`  - ${datasetType} : ${key}`);
+    const dataset = Object.fromEntries(
+      Object.entries(datasets[key as keyof typeof datasets]).sort(
+        (a, b) => a[1].order - b[1].order
+      )
+    );
+    for (const key in dataset) {
+      if (dataset[key].data) {
+        const { label, mean, standardDeviation } = dataset[key];
+        labels.push(label);
+        datalabels.push(`${mean}±${standardDeviation}`);
+        backgroundColors.push(
+          CONFIG.functions.reduce(
+            (prev, curr) =>
+              curr.functionName === key ? curr.chart.backgroundColor : prev,
+            ""
+          )
+        );
+        borderColors.push(
+          CONFIG.functions.reduce(
+            (prev, curr) =>
+              curr.functionName === key ? curr.chart.borderColor : prev,
+            ""
+          )
+        );
+        offsetData.push(mean - standardDeviation);
+        leftData.push(standardDeviation);
+        rightData.push(standardDeviation);
+        max = Math.max(max, mean + standardDeviation);
+        min = Math.min(min, mean - standardDeviation);
+      }
+    }
+    createChart({
+      backgroundColors,
+      borderColors,
+      datalabels,
+      datasetName: key,
+      datasetType,
+      gitBranch,
+      labels,
+      leftData,
+      max: Math.ceil(max / 10) * 10,
+      min: Math.floor(min / 5) * 5,
```

Comment:
add a comment documenting any magic numbers - it's not immediately clear why we are performing these computations